### ISSUE TITLE
How to use opacus if using DP for parameters

### ISSUE THREAD

**Issue by zipckkk** [pytorch/opacus#669]
In federated learning, I want to use DP for the local updates returned by the client instead of the gradients, that is, add Gaussian noise that satisfies DP to the delta weight. Similar to the following algorithm. How can I use opacus?
![image](https://github.com/user-attachments/assets/afd8ad90-97af-4d7c-9e73-73832ac9b174)

Currently, I first calculate the noise_multiplier and then add the Gaussian noise. Below is my code. Is it correct?

```
import torch
import opacus
from opacus.accountants.utils import get_noise_multiplier
from options import parse_args
from torch import autograd

args = parse_args()


def compute_noise_multiplier(target_epsilon, target_delta, global_epoch, local_epoch, batch_size, client_data_sizes):
    total_dataset_size = sum(client_data_sizes)
    sample_rate = batch_size / total_dataset_size * args.user_sample_rate
    total_steps = args.user_sample_rate * (sum([global_epoch * local_epoch * (client_data_size / batch_size) for client_data_size in client_data_sizes]))

    noise_multiplier = get_noise_multiplier(
        target_epsilon=target_epsilon,
        target_delta=target_delta,
        sample_rate=sample_rate,
        steps=total_steps, 
        accountant="rdp"
    )

    return noise_multiplier

noise_multiplier = compute_noise_multiplier(target_epsilon, target_delta, global_epoch, local_epoch, batch_size,
                                                client_data_sizes)

norm = torch.sqrt(sum([torch.sum(param ** 2) for param in client_update]))
clip_rate = max(1, (norm / clipping_bound))
clipped_update = [(param / clip_rate) for param in client_update]

noisy_updates = []
for clipped_update in clipped_updates:
        noise_stddev = torch.sqrt(torch.tensor((clipping_bound ** 2) * (noise_multiplier ** 2) / num_clients))
        noise = [torch.randn_like(param) * noise_stddev for param in clipped_update]
        noisy_update = [clipped_param + noise_param for clipped_param, noise_param in zip(clipped_update, noise)]
        noisy_updates.append(noisy_update)
```

**Comment by EnayatUllah** [comment]
I think this is indeed the right way. Does this work as intended?

An major functionality of Opacus is to calculate per-sample gradients, which is not directly exposed in PyTorch. In your case, since you already have the list of per-client updates, you can directly clip and add noise, which is what the code does. Opacus can still be used for privacy accounting.

**Comment by HuanyuZhang** [comment]
As an alternative to Opacus, you can use flower (https://flower.ai/docs/framework/how-to-use-differential-privacy.html) as a FL library.
